File: backend/mob-api/app/Http/Requests/Happening/CreateHappeningRequest.php
<?php

namespace App\Http\Requests\Happening;

use App\Enums\HappeningCategory;
use App\Enums\HappeningType;
use Illuminate\Contracts\Validation\Validator;
use Illuminate\Foundation\Http\FormRequest;
use Illuminate\Http\Exceptions\HttpResponseException;
use Illuminate\Validation\Rule;

class CreateHappeningRequest extends FormRequest
{
    public function authorize(): bool
    {
        return true;
    }

    public function rules(): array
    {
        return [
            'title' => ['required', 'string', 'max:255'],
            'description' => ['required', 'string', 'max:2000'],
            'category' => ['required', 'string', Rule::in(HappeningCategory::cases())],
            'type' => ['required', 'string', Rule::in(HappeningType::cases())],
            'latitude' => ['required', 'numeric', 'between:-90,90'],
            'longitude' => ['required', 'numeric', 'between:-180,180'],
            'radius_meters' => ['nullable', 'integer', 'min:100', 'max:5000'],
            'address' => ['required', 'string', 'max:500'],
            'starts_at' => ['nullable', 'date', 'after:now'],
            'ends_at' => ['nullable', 'date', 'after:starts_at'],
            'is_ticketed' => ['nullable', 'boolean'],
            'ticket_price' => ['required_if:is_ticketed,true', 'nullable', 'numeric', 'min:100'],
            'ticket_quantity' => ['required_if:is_ticketed,true', 'nullable', 'integer', 'min:1'],
            'snaps' => ['nullable', 'array', 'max:' . config('mob.max_snaps_per_post', 5)],
            'snaps.*.media_url' => ['required_with:snaps', 'string', 'url', 'max:2048'],
            'snaps.*.media_type' => ['required_with:snaps', 'string', 'in:image,video'],
        ];
    }

    public function withValidator(Validator $validator): void
    {
        $validator->after(function (Validator $validator) {
            $type = $this->input('type');

            if ($type === HappeningType::EVENT->value && ! $this->filled('starts_at')) {
                $validator->errors()->add('starts_at', 'The starts_at field is required for events.');
            }

            if ($this->boolean('is_ticketed') && $type !== HappeningType::EVENT->value) {
                $validator->errors()->add('is_ticketed', 'Ticketing is only allowed for events.');
            }

            if ($type === HappeningType::CASUAL->value && $this->filled('radius_meters')) {
                $snaps = $this->input('snaps');
                if (empty($snaps) || ! is_array($snaps) || count($snaps) < 1) {
                    $validator->errors()->add('snaps', 'Casual area-based happenings require at least one snap.');
                }
            }
        });
    }

    protected function failedValidation(Validator $validator): void
    {
        throw new HttpResponseException(response()->json([
            'success' => false,
            'message' => 'Validation failed',
            'errors' => $validator->errors(),
        ], 422));
    }
}

File: mob_app/lib/features/happenings/data/models/create_happening_request.dart
import 'package:flutter/foundation.dart';

import '../../../../shared/models/enums.dart';


class CreateHappeningRequest {
  const CreateHappeningRequest({
    required this.title,
    required this.category,
    required this.type,
    required this.latitude,
    required this.longitude,
    required this.address,
    this.description,
    this.radiusMeters,
    this.isHappeningNow = false,
    this.startsAt,
    this.endsAt,
    this.isTicketed = false,
    this.ticketPrice,
    this.ticketQuantity,
    this.snaps = const [],
  });

  final String title;


  final String? description;

  final HappeningCategory category;
  final HappeningType type;
  final double latitude;
  final double longitude;
  final String address;
  final int? radiusMeters;


  final bool isHappeningNow;

  final DateTime? startsAt;
  final DateTime? endsAt;
  final bool isTicketed;
  final double? ticketPrice;
  final int? ticketQuantity;


  final List<SnapPayload> snaps;


  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      'title': title,


      'description': (description != null && description!.trim().isNotEmpty)
          ? description!
          : ' ',
      'category': category.value,
      'type': type.value,
      'latitude': latitude,
      'longitude': longitude,
      'address': address,

      'is_ticketed': isTicketed,
    };

    if (radiusMeters != null) {
      json['radius_meters'] = radiusMeters;
    }


    if (isHappeningNow) {
      json['starts_at'] = DateTime.now()
          .add(const Duration(seconds: 30))
          .toIso8601String();
    } else if (startsAt != null) {
      json['starts_at'] = startsAt!.toIso8601String();
    }

    debugPrint(
      '[Mob/CreateRequest] isHappeningNow=$isHappeningNow '
      'startsAt=$startsAt '
      'json starts_at=${json['starts_at']}',
    );

    if (endsAt != null) {
      json['ends_at'] = endsAt!.toIso8601String();
    }

    if (isTicketed) {
      if (ticketPrice != null) json['ticket_price'] = ticketPrice;
      if (ticketQuantity != null) json['ticket_quantity'] = ticketQuantity;
    }

    if (snaps.isNotEmpty) {
      json['snaps'] = snaps.map((s) => s.toJson()).toList();
    }

    return json;
  }
}


class SnapPayload {
  const SnapPayload({
    required this.mediaUrl,
    required this.mediaType,
  });


  final String mediaUrl;


  final String mediaType;

  Map<String, dynamic> toJson() => {
        'media_url': mediaUrl,
        'media_type': mediaType,
      };
}
